File: 54-spiral-matrix/spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    const result = []
    let queue = [[0,0, "R"]], n = matrix.length, m = matrix[0].length

    while (queue.length) {
        let [r, c, v] = queue.shift();
        if (r < 0 || c < 0 || r >= n || c >= m) break
        if (matrix[r][c] === "-") break;
        result.push(matrix[r][c])
        matrix[r][c] = "-"
        if (v === "R") {
            while (c+1 < m && matrix[r][c+1] !== "-") {
                c++;
                result.push(matrix[r][c])
                matrix[r][c] = "-"
            }
            queue.push([r+1, c, "D"])
        } else if (v === "L") {
            while (c-1 >= 0 && matrix[r][c-1] !== "-") {
                c--;
                result.push(matrix[r][c])
                matrix[r][c] = "-";
            }
            queue.push([r-1, c, "U"])

        } else if (v === "D") {
            while (r+1 < n && matrix[r+1][c] !== "-") {
                r++;
                result.push(matrix[r][c])
                matrix[r][c] = "-";
            }
            queue.push([r, c-1, "L"])

        } else if (v === "U") {
            while (r-1 >= 0 && matrix[r-1][c] !== "-") {
                r--;
                result.push(matrix[r][c])
                matrix[r][c] = "-"
            }
            queue.push([r, c+1, "R"])
        }

    }
    return result
};

File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();
        int r = 0; int c = 0; int n = matrix.length; int m = matrix[0].length;

        while (inRange(r,c, n, m) && matrix[r][c] != Integer.MAX_VALUE) {

                // right
                while (inRange(r,c, n, m) && matrix[r][c] != Integer.MAX_VALUE) {
                    result.add(matrix[r][c]);
                    matrix[r][c] = Integer.MAX_VALUE;
                    if (c+1 < m && matrix[r][c+1] != Integer.MAX_VALUE) {
                        c++;
                    }
                }
                r++;
                // down
                while (inRange(r,c, n, m) && matrix[r][c] != Integer.MAX_VALUE) {
                    result.add(matrix[r][c]);
                    matrix[r][c] = Integer.MAX_VALUE;
                    if (r+1 < n && matrix[r+1][c] != Integer.MAX_VALUE) {
                        r++;
                    }
                }
                c--;
                // left
                while (inRange(r,c, n, m) && matrix[r][c] != Integer.MAX_VALUE) {
                    result.add(matrix[r][c]);
                    matrix[r][c] = Integer.MAX_VALUE;
                    if (c-1 >= 0 && matrix[r][c-1] != Integer.MAX_VALUE) {
                        c--;
                    }
                }
                r--;
                // up
                while (inRange(r,c, n, m) && matrix[r][c] != Integer.MAX_VALUE) {
                    result.add(matrix[r][c]);
                    matrix[r][c] = Integer.MAX_VALUE;
                    if (r-1 >= 0 && matrix[r-1][c] != Integer.MAX_VALUE) {
                        r--;
                    }
                }
                c++;
            }
        return result;
    }

    private boolean inRange (int r, int c, int n, int  m) {
        return r >= 0 && c >= 0 && r < n && c < m;
    }
}